### PR TITLE
DRAFT: [BUG] Transferring ownership does not emit an AccountUpdated event

### PR DESCRIPTION
**Pull request by web3ashlee**
Resolves: #67

### CODE REVIEW

**Comment by sisyphusSmiling** [comment]
Thanks for the PR @web3ashlee! Taking a look, I'm getting the sense we're trying to pack too much into the `AccountUpdated` event.

So far, `AccountUpdated` is emitted whenever an `ChildAccount` or `ProxyAccount` Capability is added or removed to a `Manager`. Granting ownership feels like a distinct action with broader implications for grantor is recipient of that ownership, warranting its own event IMO.

Reflecting back on the issue, I'm not convinced this change is necessary and in fact makes is more difficult to infer state from watching events. Totally open to being wrong, but wondering what others think here.

File: contracts/HybridCustody.cdc [hunk 1]
```diff
@@ -732,7 +736,9 @@ pub contract HybridCustody {
             self.acctOwner = to
             self.relinquishedOwnership = false
 
-            emit OwnershipGranted(id: self.uuid, child: self.acct.address, owner: to)
+            emit OwnershipUpdated(id: self.uuid, child: self.acct.address, owner: to, active: false)
```
Comment (sisyphusSmiling) [inline contracts/HybridCustody.cdc:739]: The concern I have here is that `giveOwnership()` == `removeOwned()` whereas before these actions were distinct.

File: contracts/HybridCustody.cdc [hunk 2]
```diff
@@ -732,7 +736,9 @@ pub contract HybridCustody {
...
+            //TODO do we need both events
+            emit AccountUpdated(id: self.uuid, child: self.acct.address, parent: to, proxy: false, active: false)
```
Comment (sisyphusSmiling) [inline contracts/HybridCustody.cdc:741]: Whichever path we choose, we won't need a second event. An event will be emitted by the user when they add the published Capability to their manager with `addOwnedAccount()`

**Review by sisyphusSmiling — COMMENTED** [review]
I think the confusion is around the fact that `AccountUpdated` was intended to notify that an account Capability was added to a `Manager`, not that the state of a Proxy/ChildAccount changed. We're now trying to accomplish both with `OwnershipUpdated` and `AccountUpdated` - when caps are added to a `Manager` **_and_** when a `ChildAccount` gives ownership.

Thinking through this, here's what the values look like in the pre-existing state:
> Let's say the child address is 0x07 and parent 0x13

|Action|Event & Values|
|---|---|
|ProxyAccount Cap added to a Manager|`AccountUpdated(id: xx, child: 0x07, parent: 0x13, proxy: true, active: true)`|
|ProxyAccount Cap removed from a Manager|`AccountUpdated(id: xx, child: 0x07, parent: 0x13, proxy: true, active: false)`|
|ChildAccount Cap added to a Manager (ownership claimed)|`AccountUpdated(id: xx, child: 0x07, parent: 0x13, proxy: false, active: true)`|
|ChildAccount Cap removed from a Manager|`AccountUpdated(id: xx, child: 0x07, parent: 0x13, proxy: false, active: false)`|
|ChildAccount grant ownership|`OwnershipGranted(id: xx, child: 0x07, owner: 0x13)`|

Again, all `AccountUpdated` events relate to the user's account, where the `Manager` lives which would be relevant for a wallet provider or an application facilitating linking to a user's account or an indexer identifying linked accounts - we can follow a path from parent to owned/child account by listening to all of these events as they're distinct actions.

Splitting along type of encapsulating resource - Proxy v. ChildAccount - on Account v. OwnershipUpdated makes this more ambiguous.

Let's try the same truth table:
> Reminder, child address is 0x07 and parent 0x13. Also removing `proxy: Bool` field makes sense here.

|Action|Event & Values|
|---|---|
|ProxyAccount Cap added to a Manager|`AccountUpdated(id: xx, child: 0x07, parent: 0x13, active: true)`|
|ProxyAccount Cap removed from a Manager|`AccountUpdated(id: xx, child: 0x07, parent: 0x13, active: false)`|
|ChildAccount Cap added to a Manager (ownership claimed)|`OwnershipUpdated(id: xx, child: 0x07, parent: 0x13, active: true)`|
|ChildAccount Cap removed from a Manager|`OwnershipUpdated(id: xx, child: 0x07, parent: 0x13, active: false)`|
|ChildAccount grant ownership|`OwnershipGranted(id: xx, child: 0x07, owner: 0x13)`|

The event `OwnershipUpdated(id: xx, child: 0x07, parent: 0x13, active: true)` now has two meanings - `Manager.removeOwned()` & `ChildAccount.giveOwnership()`. One is relevant to a user and another relevant to the existing owner and those two can be entirely different entities.

That's all to say that I still think the existing events satisfy everything we need to know. `OwnershipGranted` let's someone know when they can claim ownership of a `ChildAccount` and `AccountUpdated` informs them whenever a Proxy/ChildAccount cap has been added/removed from their `Manager`.

If consensus demands we alter the current event structure, I'd say we remove the `proxy: Bool` field from `AccountUpdated`, add `OwnershipUpdated` and keep the `OwnershipGranted` event - but this would mean adding an event instead of consolidating any. If keeping `OwnershipGranted` just isn't an option, I'd prefer to get rid of it to avoid ambiguity.

**Comment by bjartek** [comment]
> 

I think this makes sense. Love the tables. You could also argue that not using proxy field and having two different events could make sense, allthough I think having it all in one makes it easier to track those event changes.

**Comment by sisyphusSmiling** [comment]
Thanks for the feedback @bjartek! To clarify,

>  I think having it all in one makes it easier to track those event changes.

Do you mean here that you prefer the top or the bottom table, or some other alternative?

File: contracts/HybridCustody.cdc [hunk 3]
```diff
@@ -55,6 +55,7 @@ pub contract HybridCustody {
         pub fun isChildOf(_ addr: Address): Bool
         pub fun getParentsAddresses(): [Address]
         pub fun borrowAccount(): &AuthAccount?
+        pub fun getOwner(): Address?
```
Comment (sisyphusSmiling) [inline contracts/HybridCustody.cdc:58]: Perhaps less important, but I think we should also add `getPendingOwner()` here. What do you think?

**Review by sisyphusSmiling — COMMENTED** [review]
Looks like this branch needs to pull the changes from #86. I believe the changes there would affect how the inline calls are handled within the events due to owner assignment on claiming.

**Comment by austinkline** [comment]
> Looks like this branch needs to pull the changes from #86. I believe the changes there would affect how the inline calls are handled within the events due to owner assignment on claiming.

Rebased!

**Review by sisyphusSmiling — APPROVED** [review]
LGTM! 🚀

**Pull request merged**